include/runtime/collections/rangemap.h

```diff
@@ -0,0 +1,380 @@
+        // Adjust the bounds of the inserted range as needed, and delete
+        // the relevant range.
+        is = rrs < is ? rrs : is;
+        ie = rre > ie ? rre : ie;
+        tmpmap = tmpmap.deleted(rr);
+      } else {
+        // The inserted value is different from the value stored
+        // in the relevant range.
+        if (ranges_overlap(r, rr)) {
+          // The ranges overlap.
+          // Delete the relevant range, and insert up to two ranges to
+          // the left/right of the inserted range, as needed to
+          // account for the bounds of the relevant range.
+          tmpmap = tmpmap.deleted(rr);
+          if (rrs < is) {
+            tmpmap = tmpmap.inserted(std::make_pair(rrs, is), rv);
+          }
+          if (rre > ie) {
+            tmpmap = tmpmap.inserted(std::make_pair(ie, rre), rv);
+          }
+        } else {
+          // The ranges do not overlap. NOOP.
+        }
+      }
+    }
+
+    // Finally, insert a range with the properly adjusted bounds and the
+    // new value. Create a new range map.
+    tmpmap = tmpmap.inserted(std::make_pair(is, ie), v);
+    return RangeMap(tmpmap);
+  }
+
+  /* Given a range to be removed and a map, return a map that does not
+     * contain the removed range or any part of it.
+     * If the range to be removed partially overlaps with any ranges in the
+     * map, then the boundaries of these ranges are adjusted in the resulting
+     * map so that they do not overlap with the removed range.
+     */
+  RangeMap deleted(Range r) {
+    // Empty ranges do not make sense here.
+    if (range_is_empty(r))
+      CONSTRUCT_MSG_AND_THROW("Delete empty range from range map");
+
+    std::vector<std::pair<Range, V>> ranges;
+    getDeletionRelevantRanges(_map, r, ranges);
+    // Each relevant range may lead to changes to the existing underlying
+    // treemap data structure.
+    // We iterate over the collected relevant ranges to collect and apply
+    // these changes.
+    T ds = r.first;
+    T de = r.second;
+    RBTree<Range, V> tmpmap = _map;
+    for (auto &p : ranges) {
+      Range rr = p.first;
+      V rv = p.second;
+      assert(ranges_overlap(r, rr));
+      T rrs = rr.first;
+      T rre = rr.second;
+      tmpmap = tmpmap.deleted(rr);
+      if (rrs < ds) {
+        tmpmap = tmpmap.inserted(std::make_pair(rrs, ds), rv);
+      }
+      if (rre > de) {
+        tmpmap = tmpmap.inserted(std::make_pair(de, rre), rv);
+      }
+    }
+    return RangeMap(tmpmap);
+  }
+
+  // Return true if range r partially or completely overlaps with any key range
+  // stored in the map
+  bool overlaps(Range r) const { return overlaps(_map, r); }
+
+  void print() {
+    std::cout << "--------------------------" << std::endl;
+    forEach(_map, [](Range x, V v) {
+      std::cout << "[ " << x.first << ".." << x.second << " ) -> " << v
+                << std::endl;
+    });
+    std::cout << "--------------------------" << std::endl;
+  }
```

Comment:
If we're exposing `print()` in the public namespace for debugging, let's add an argument so that it doesn't always have to go to `std::cout`.

Reply:
Addressed.